File: sdk-typescript/packages/fleet/src/fleet.ts
/**
 * WAVE SDK - Fleet API
 *
 * Device fleet management for Desktop Nodes. Register, monitor, and control
 * nodes in your organization's streaming infrastructure.
 *
 * NOTE: This is a client SDK. All authorization checks are performed server-side.
 * The API will return 403 Forbidden if the user lacks required permissions.
 */

import type {
  WaveClient,
  PaginationParams,
  PaginatedResponse,
  Timestamps,
  Metadata,
} from '@wave-av/core';

// ============================================================================
// Types
// ============================================================================

/**
 * Node connection status
 */
export type NodeStatus = "online" | "offline" | "maintenance" | "updating";

/**
 * Node health state
 */
export type NodeHealth = "healthy" | "degraded" | "critical";

/**
 * Fleet node object
 */
export interface FleetNode extends Timestamps {
  id: string;
  organization_id: string;
  name: string;
  status: NodeStatus;
  health: NodeHealth;
  ip_address: string;
  version: string;
  os: string;
  cpu_usage: number;
  memory_usage: number;
  gpu_usage: number;
  device_count: number;
  last_seen_at: string;
  registered_at: string;
  tags: string[];
  metadata: Metadata;
}

/**
 * Device attached to a node
 */
export interface NodeDevice {
  id: string;
  node_id: string;
  name: string;
  type: "camera" | "microphone" | "display" | "capture_card" | "ndi_source";
  status: string;
  driver_version: string;
}

/**
 * Register a new node
 */
export interface RegisterNodeRequest {
  name: string;
  os: string;
  version: string;
  tags?: string[];
  metadata?: Metadata;
}

/**
 * Update a node
 */
export interface UpdateNodeRequest {
  name?: string;
  tags?: string[];
  metadata?: Metadata;
}

/**
 * List nodes filters
 */
export interface ListNodesParams extends PaginationParams {
  status?: NodeStatus;
  health?: NodeHealth;
  os?: string;
  order_by?: "name" | "created_at" | "last_seen_at" | "cpu_usage";
  order?: "asc" | "desc";
}

/**
 * Command to send to a node
 */
export interface NodeCommand {
  type: "restart" | "update" | "shutdown" | "scan_devices" | "clear_cache";
  params?: Record<string, unknown>;
}

/**
 * Node resource metrics snapshot
 */
export interface NodeMetrics {
  node_id: string;
  cpu_usage: number;
  memory_usage: number;
  gpu_usage: number;
  network_in_mbps: number;
  network_out_mbps: number;
  disk_usage: number;
  uptime_seconds: number;
  device_count: number;
  active_bridges: number;
  timestamp: string;
}

// ============================================================================
// Fleet API
// ============================================================================

/**
 * Fleet API client
 *
 * All operations require appropriate permissions. Authorization is enforced
 * server-side - the API returns 403 if the authenticated user lacks access.
 *
 * @example
 * ```typescript
 * import { WaveClient } from '@wave-av/core';
 * import { FleetAPI } from '@wave-av/fleet';
 *
 * const client = new WaveClient({ apiKey: 'your-api-key' });
 * const fleet = new FleetAPI(client);
 *
 * // List all online nodes
 * const nodes = await fleet.list({ status: 'online' });
 *
 * // Register a new node
 * const node = await fleet.register({
 *   name: 'Studio-Node-01',
 *   os: 'linux',
 *   version: '2.4.0',
 *   tags: ['studio-a'],
 * });
 *
 * // Send a command to scan for devices
 * await fleet.sendCommand(node.id, { type: 'scan_devices' });
 * ```
 */
export class FleetAPI {
  private readonly client: WaveClient;
  private readonly basePath = "/v1/fleet/nodes";

  constructor(client: WaveClient) {
    this.client = client;
  }

  /**
   * List fleet nodes with optional filters
   *
   * Requires: fleet:read permission
   */
  async list(params?: ListNodesParams): Promise<PaginatedResponse<FleetNode>> {
    const queryParams: Record<string, string | number | boolean | undefined> = {
      limit: params?.limit,
      offset: params?.offset,
      cursor: params?.cursor,
      status: params?.status,
      health: params?.health,
      os: params?.os,
      order_by: params?.order_by,
      order: params?.order,
    };

    return this.client.get<PaginatedResponse<FleetNode>>(this.basePath, {
      params: queryParams,
    });
  }

  /**
   * Get a node by ID
   *
   * Requires: fleet:read permission
   */
  async get(nodeId: string): Promise<FleetNode> {
    return this.client.get<FleetNode>(`${this.basePath}/${nodeId}`);
  }

  /**
   * Register a new node
   *
   * Requires: fleet:create permission
   */
  async register(request: RegisterNodeRequest): Promise<FleetNode> {
    return this.client.post<FleetNode>(this.basePath, request);
  }

  /**
   * Update a node
   *
   * Requires: fleet:update permission
   */
  async update(nodeId: string, request: UpdateNodeRequest): Promise<FleetNode> {
    return this.client.patch<FleetNode>(`${this.basePath}/${nodeId}`, request);
  }

  /**
   * Deregister (remove) a node
   *
   * Requires: fleet:remove permission (server-side RBAC enforced)
   */
  async deregister(nodeId: string): Promise<void> {
    await this.client.delete(`${this.basePath}/${nodeId}`);
  }

  /**
   * Get current health status of a node
   *
   * Requires: fleet:read permission
   */
  async getHealth(
    nodeId: string,
  ): Promise<{ health: NodeHealth; details: Record<string, unknown> }> {
    return this.client.get<{ health: NodeHealth; details: Record<string, unknown> }>(
      `${this.basePath}/${nodeId}/health`,
    );
  }

  /**
   * List devices attached to a node
   *
   * Requires: fleet:read permission
   */
  async listDevices(nodeId: string): Promise<NodeDevice[]> {
    return this.client.get<NodeDevice[]>(`${this.basePath}/${nodeId}/devices`);
  }

  /**
   * Send a command to a node
   *
   * Requires: fleet:command permission
   */
  async sendCommand(
    nodeId: string,
    command: NodeCommand,
  ): Promise<{ command_id: string; status: string }> {
    return this.client.post<{ command_id: string; status: string }>(
      `${this.basePath}/${nodeId}/commands`,
      command,
    );
  }

  /**
   * Get current resource metrics for a node
   *
   * Requires: fleet:read permission
   */
  async getMetrics(nodeId: string): Promise<NodeMetrics> {
    return this.client.get<NodeMetrics>(`${this.basePath}/${nodeId}/metrics`);
  }

  /**
   * Wait for a node to come online
   */
  async waitForOnline(
    nodeId: string,
    options?: {
      pollInterval?: number;
      timeout?: number;
      onProgress?: (node: FleetNode) => void;
    },
  ): Promise<FleetNode> {
    const pollInterval = options?.pollInterval || 5000;
    const timeout = options?.timeout || 120000; // 2 minutes default
    const startTime = Date.now();

    while (Date.now() - startTime < timeout) {
      const node = await this.get(nodeId);

      if (options?.onProgress) {
        options.onProgress(node);
      }

      if (node.status === "online") {
        return node;
      }

      await new Promise((resolve) => setTimeout(resolve, pollInterval));
    }

    throw new Error(`Node ${nodeId} did not come online within ${timeout}ms`);
  }
}

/**
 * Create a Fleet API instance
 */
export function createFleetAPI(client: WaveClient): FleetAPI {
  return new FleetAPI(client);
}

File: sdk-python/wave/fleet.py
"""WAVE SDK - Fleet API. Desktop Node device fleet management."""
from __future__ import annotations
import time
from typing import Any
from pydantic import BaseModel
from wave.client import WaveClient

class FleetNode(BaseModel):
    id: str; organization_id: str; name: str; status: str; health: str; ip_address: str | None = None; version: str | None = None; os: str | None = None; cpu_usage: float = 0; memory_usage: float = 0; device_count: int = 0; last_seen_at: str | None = None; created_at: str; updated_at: str

class NodeDevice(BaseModel):
    id: str; node_id: str; name: str; type: str; status: str; driver_version: str | None = None

class FleetAPI:
    def __init__(self, client: WaveClient): self._client = client; self._base = "/v1/fleet/nodes"
    def list(self, **params: Any) -> dict: return self._client.get(self._base, params={k: v for k, v in params.items() if v is not None})
    def get(self, node_id: str) -> FleetNode: return FleetNode(**self._client.get(f"{self._base}/{node_id}"))
    def register(self, name: str, os: str, version: str, **kwargs: Any) -> FleetNode: return FleetNode(**self._client.post(self._base, json={"name": name, "os": os, "version": version, **kwargs}))
    def update(self, node_id: str, **kwargs: Any) -> FleetNode: return FleetNode(**self._client.patch(f"{self._base}/{node_id}", json=kwargs))
    def deregister(self, node_id: str) -> None: self._client.delete(f"{self._base}/{node_id}")
    def get_health(self, node_id: str) -> dict: return self._client.get(f"{self._base}/{node_id}/health")
    def list_devices(self, node_id: str) -> list[NodeDevice]: return [NodeDevice(**d) for d in self._client.get(f"{self._base}/{node_id}/devices")]
    def send_command(self, node_id: str, command_type: str, params: dict | None = None) -> dict: return self._client.post(f"{self._base}/{node_id}/commands", json={"type": command_type, "params": params})
    def get_metrics(self, node_id: str) -> dict: return self._client.get(f"{self._base}/{node_id}/metrics")
    def wait_for_online(self, node_id: str, poll_interval: float = 5.0, timeout: float = 120.0) -> FleetNode:
        start = time.time()
        while time.time() - start < timeout:
            n = self.get(node_id)
            if n.status == "online": return n
            time.sleep(poll_interval)
        raise TimeoutError(f"Node did not come online within {timeout}s")
